File: android/src/main/java/org/ligi/passandroid/ui/UnzipPassController.java
package org.ligi.passandroid.ui;

import android.content.Context;

import net.lingala.zip4j.core.ZipFile;
import net.lingala.zip4j.exception.ZipException;

import org.json.JSONException;
import org.json.JSONObject;
import org.ligi.axt.AXT;
import org.ligi.passandroid.App;
import org.ligi.passandroid.helper.SafeJSONReader;
import org.ligi.passandroid.model.InputStreamWithSource;

import java.io.File;
import java.io.FileNotFoundException;
import java.util.UUID;

public class UnzipPassController {

    public interface SuccessCallback {
        void call(String uuid);
    }

    public interface FailCallback {
        void fail(String reason);
    }

    public static void processInputStream(InputStreamUnzipControllerSpec spec) {
        try {
            final File tempFile = File.createTempFile("ins", "pass");
            AXT.at(spec.inputStreamWithSource.getInputStream()).toFile(tempFile);
            processFile(new FileUnzipControllerSpec(tempFile.getAbsolutePath(), spec));
            tempFile.delete();
        } catch (Exception e) {
            App.component().tracker().trackException("problem processing InputStream", e, false);
            spec.failCallback.fail("problem with temp file" + e);
        }
    }

    public static void processFile(FileUnzipControllerSpec spec) {

        String uuid = UUID.randomUUID().toString();
        String path = spec.context.getCacheDir() + "/temp/" + uuid + "/";

        final File dir_file = new File(path);
        dir_file.mkdirs();

        if (!dir_file.exists()) {
            spec.failCallback.fail("Problem creating the temp dir: " + path);
            return;
        }

        AXT.at(new File(path + "source.obj")).writeObject(spec.source);

        try {
            final ZipFile zipFile = new ZipFile(spec.zipFileString);
            zipFile.extractAll(path);
        } catch (ZipException e) {
            e.printStackTrace();
        }


        JSONObject manifest_json;
        try {
            final String readToString = AXT.at(new File(path + "/manifest.json")).readToString();
            manifest_json = SafeJSONReader.readJSONSafely(readToString);
        } catch (FileNotFoundException e) {
            spec.failCallback.fail("Pass contains no Manifest - or PassAndroid could not read it");
            return;
        } catch (Exception e) {
            spec.failCallback.fail("Problem with manifest.json: " + e);
            return;
        }


        try {
            uuid = manifest_json.getString("pass.json");
            final String rename_str = spec.targetPath + "/" + uuid;
            new File(spec.targetPath).mkdirs();
            final File rename_file = new File(rename_str);

            if (spec.overwrite && rename_file.exists()) {
                AXT.at(rename_file).deleteRecursive();
            }

            if (!rename_file.exists()) {
                new File(path + "/").renameTo(rename_file);
            } else {
                spec.failCallback.fail("Pass with same ID exists");
                return;
            }

        } catch (JSONException e) {
            spec.failCallback.fail("Problem with pass.json: " + e);
            return;
        }

        spec.onSuccessCallback.call(uuid);

    }

    public static class UnzipControllerSpec {
        public final Context context;
        public final SuccessCallback onSuccessCallback;
        public final FailCallback failCallback;
        public String targetPath;
        public boolean overwrite = false;

        public UnzipControllerSpec(String targetPath, Context context, SuccessCallback onSuccessCallback, FailCallback failCallback) {
            this.context = context;
            this.onSuccessCallback = onSuccessCallback;
            this.failCallback = failCallback;
            this.targetPath = targetPath;
        }

        public UnzipControllerSpec(Context context, SuccessCallback onSuccessCallback, FailCallback failCallback) {
            this(App.component().settings().getPassesDir(), context, onSuccessCallback, failCallback);
        }


    }

    public static class FileUnzipControllerSpec extends UnzipControllerSpec {
        public final String zipFileString;
        public final String source;

        public FileUnzipControllerSpec(final String fileName,
                                       final InputStreamUnzipControllerSpec spec) {
            super(spec.targetPath, spec.context, spec.onSuccessCallback, spec.failCallback);
            zipFileString = fileName;
            source = spec.inputStreamWithSource.getSource();
            overwrite = spec.overwrite;

        }
    }

    public static class InputStreamUnzipControllerSpec extends UnzipControllerSpec {
        final InputStreamWithSource inputStreamWithSource;

        public InputStreamUnzipControllerSpec(final InputStreamWithSource inputStreamWithSource,
                                              final Context context,
                                              final SuccessCallback onSuccessCallback,
                                              final FailCallback failCallback) {
            super(context, onSuccessCallback, failCallback);
            this.inputStreamWithSource = inputStreamWithSource;
        }


    }

}

File: android/src/main/java/org/ligi/passandroid/ui/UnzipPassController.kt
package org.ligi.passandroid.ui

import android.content.Context
import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.graphics.pdf.PdfRenderer
import android.os.Build
import android.os.ParcelFileDescriptor
import net.lingala.zip4j.ZipFile
import net.lingala.zip4j.exception.ZipException
import okio.buffer
import okio.source
import org.json.JSONObject
import org.koin.core.component.KoinComponent
import org.koin.core.component.inject
import org.ligi.passandroid.Tracker
import org.ligi.passandroid.functions.createPassForImageImport
import org.ligi.passandroid.functions.createPassForPDFImport
import org.ligi.passandroid.functions.readJSONSafely
import org.ligi.passandroid.model.InputStreamWithSource
import org.ligi.passandroid.model.PassStore
import org.ligi.passandroid.model.Settings
import timber.log.Timber
import java.io.File
import java.io.FileOutputStream
import java.util.*

object UnzipPassController : KoinComponent {

    val tracker :Tracker by inject()
    val settings : Settings by inject()

    interface SuccessCallback {
        fun call(uuid: String)
    }

    interface FailCallback {
        fun fail(reason: String)
    }

    fun processInputStream(spec: InputStreamUnzipControllerSpec) {
        try {
            val tempFile = File.createTempFile("ins", "pass")
            spec.inputStreamWithSource.inputStream.copyTo(FileOutputStream(tempFile))
            processFile(FileUnzipControllerSpec(tempFile.absolutePath, spec))
            tempFile.delete()
        } catch (e: Exception) {
            tracker.trackException("problem processing InputStream", e, false)
            spec.failCallback?.fail("problem with temp file: $e")
        }

    }

    private fun processFile(spec: FileUnzipControllerSpec) {

        var uuid = UUID.randomUUID().toString()
        val path = File(spec.context.cacheDir, "temp/$uuid")

        path.mkdirs()

        if (!path.exists()) {
            spec.failCallback?.fail("Problem creating the temp dir: $path")
            return
        }

        File(path, "source.obj").bufferedWriter().write(spec.source)

        try {
            val zipFile = ZipFile(spec.zipFileString)
            zipFile.extractAll(path.absolutePath)
        } catch (e: ZipException) {
            e.printStackTrace()
        }


        val manifestFile = File(path, "manifest.json")
        val espassFile = File(path, "main.json")
        val manifestJSON: JSONObject

        when {
            manifestFile.exists() -> try {
                val readToString = manifestFile.bufferedReader().readText()
                manifestJSON = readJSONSafely(readToString)!!
                uuid = manifestJSON.getString("pass.json")
            } catch (e: Exception) {
                spec.failCallback?.fail("Problem with manifest.json: $e")
                return
            }
            espassFile.exists() -> try {
                val readToString = espassFile.bufferedReader().readText()
                manifestJSON = readJSONSafely(readToString)!!
                uuid = manifestJSON.getString("id")
            } catch (e: Exception) {
                spec.failCallback?.fail("Problem with manifest.json: $e")
                return
            }
            else -> {
                val bitmap = BitmapFactory.decodeFile(spec.zipFileString)
                val resources = spec.context.resources

                if (bitmap != null) {
                    val imagePass = createPassForImageImport(resources)
                    val pathForID = spec.passStore.getPathForID(imagePass.id)
                    pathForID.mkdirs()

                    File(spec.zipFileString).copyTo(File(pathForID, "strip.png"))

                    spec.passStore.save(imagePass)
                    spec.passStore.classifier.moveToTopic(imagePass, "new")
                    spec.onSuccessCallback?.call(imagePass.id)
                    return
                }

                if (Build.VERSION.SDK_INT >= 21) {
                    try {
                        val file = File(spec.zipFileString)
                        val readUtf8 = file.source().buffer().readUtf8(4)
                        if (readUtf8 == "%PDF") {
                            val open = ParcelFileDescriptor.open(file, ParcelFileDescriptor.MODE_READ_ONLY)
                            val pdfRenderer = PdfRenderer(open)

                            val page = pdfRenderer.openPage(0)
                            val ratio = page.height.toFloat() / page.width

                            val widthPixels = resources.displayMetrics.widthPixels
                            val createBitmap = Bitmap.createBitmap(widthPixels, (widthPixels * ratio).toInt(), Bitmap.Config.ARGB_8888)
                            page.render(createBitmap, null, null, PdfRenderer.Page.RENDER_MODE_FOR_DISPLAY)

                            val imagePass = createPassForPDFImport(resources)
                            val pathForID = spec.passStore.getPathForID(imagePass.id)
                            pathForID.mkdirs()

                            createBitmap.compress(Bitmap.CompressFormat.PNG, 100, FileOutputStream(File(pathForID, "strip.png")))

                            spec.passStore.save(imagePass)
                            spec.passStore.classifier.moveToTopic(imagePass, "new")
                            spec.onSuccessCallback?.call(imagePass.id)
                            return
                        }
                    } catch (e: Exception) {
                        e.printStackTrace()
                    }
                }

                spec.failCallback?.fail("Pass is not espass or pkpass format :-(")
                return
            }
        }

        spec.targetPath.mkdirs()
        val renamedFile = File(spec.targetPath, uuid)

        if (spec.overwrite && renamedFile.exists()) {
            renamedFile.deleteRecursively()
        }

        if (!renamedFile.exists()) {
            path.renameTo(renamedFile)
        } else {
            Timber.i("Pass with same ID exists")
        }

        spec.onSuccessCallback?.call(uuid)
    }

    class InputStreamUnzipControllerSpec(internal val inputStreamWithSource: InputStreamWithSource, context: Context, passStore: PassStore,
                                         onSuccessCallback: SuccessCallback?, failCallback: FailCallback?) : UnzipControllerSpec(context, passStore, onSuccessCallback, failCallback, settings)

}
